File: inst/python/_jenks.py
# https://github.com/llimllib/jenks-python

import json
from pprint import pprint as pp

inf = float('inf')

def jenks_init_matrices(data, n_classes):
    #where N = len(data) and K = len(n_classes)+1
    k = n_classes + 1
    #An N x K matrix of optimal lower class limits
    lower_class_limits =    [[0.] * k for i in range(len(data))]
    #An N x K matrix of optimal variance combinations
    #the first row is 0s, the rest of the rows are [0, inf*K]
    variance_combinations = [[0.] * k] + \
                            [[0.] + [inf] * k
                             for i in range(len(data))]
    #the original code sets lc[0][i] to 1... figure out what this row is actually for!
    for i in range(1, n_classes+1):
        lower_class_limits[1][i] = 1.
    return lower_class_limits, variance_combinations




def jenks_matrices(data, n_classes):
    lower_class_limits, variance_combinations = jenks_init_matrices(data, n_classes)
    variance = 0.0
    for l in range(1, len(data)):
        sum = 0.0
        sum_squares = 0.0
        w = 0.0
        for m in range(l):
            # `III` originally
            lower_class_limit = l - m + 1
            val = data[lower_class_limit-1]
            # here we're estimating variance for each potential classing
            # of the data, for each potential number of classes. `w`
            # is the number of data points considered so far.
            w += 1
            # increase the current sum and sum-of-squares
            sum += val
            sum_squares += val * val
            # the variance at this point in the sequence is the difference
            # between the sum of squares and the total x 2, over the number
            # of samples.
            variance = sum_squares - (sum * sum) / w
            i4 = lower_class_limit - 1
            if i4 != 0:
                for j in range(2, n_classes+1):
                    if variance_combinations[l][j] >= (variance + variance_combinations[i4][j - 1]):
                        lower_class_limits[l][j] = lower_class_limit
                        variance_combinations[l][j] = variance + variance_combinations[i4][j - 1]
        lower_class_limits[l][1] = 1.
        variance_combinations[l][1] = variance
    return lower_class_limits, variance_combinations

def get_jenks_breaks(data, lower_class_limits, n_classes):
    k = len(data) - 1
    kclass = [0.] * (n_classes+1)
    countNum = n_classes
    kclass[0] = data[0]
    kclass[n_classes] = data[len(data) - 1]
    while countNum > 1:
        elt = int(lower_class_limits[k][countNum] - 2)
        kclass[countNum - 1] = data[elt]
        k = int(lower_class_limits[k][countNum] - 1)
        countNum -= 1
    return kclass

def separate_data(data, boundaries):
    clusters = []
    cluster = []
    boundaries.pop(0)
    i=0
    for d in data:
        if d <= boundaries[i]:
           cluster.append(d)
        else:
            clusters.append(cluster)
            cluster = [d]
            i += 1
    clusters.append(cluster)
    return clusters

def jenks(data, n_classes):
    if n_classes > len(data): return ###nothing?
    data.sort()
    lower_class_limits, _ = jenks_matrices(data, n_classes)
    boundaries = get_jenks_breaks(data, lower_class_limits, n_classes)
    return separate_data(data, boundaries), boundaries


def main():
    pp(jenks([1,2,3,4,5,6,7,8,9,10], 3))

File: src/_jenks.cpp
//
// Created by Colin Small on 8/2/21.
//

#include <Rcpp.h>
using namespace Rcpp;

#include "_jenks.h"
#include <vector>
#include <algorithm>

std::tuple<std::vector<std::vector<double>>, std::vector<std::vector<double>>> jenksInitMatrices(std::vector<double> data, int nClasses){
    int k = nClasses + 1;
    std::vector<std::vector<double>> lowerClassLimits(data.size(), std::vector<double>(k, 0.0));
    std::vector<std::vector<double>> varianceCombinations(data.size(), std::vector<double>(k, std::numeric_limits<double>::infinity()));
    varianceCombinations[0] = std::vector<double>(k, 0.0);

    for(int i = 1; i < nClasses+1; i++){
        lowerClassLimits[1][i] = 1;
    }

    return std::tuple<std::vector<std::vector<double>>, std::vector<std::vector<double>>>{lowerClassLimits, varianceCombinations};
}

std::tuple<std::vector<std::vector<double>>, std::vector<std::vector<double>>> jenksMatrices(std::vector<double> data, int nClasses){
    std::tuple<std::vector<std::vector<double>>, std::vector<std::vector<double>>> results = jenksInitMatrices(data, nClasses);
    std::vector<std::vector<double>> lowerClassLimits = std::get<0>(results);
    std::vector<std::vector<double>> varianceCombinations = std::get<1>(results);

    double variance = 0.0;

    for(int l = 1; l < data.size(); l++){
        double sum = 0.0;
        double sumSquares = 0.0;
        double w = 0.0;

        for(int m = 0; m < l; m++){
            int lowerClassLimit = l-m+1;
            double val = data[lowerClassLimit-1];
            w += 1;
            sum += val;
            sumSquares += val * val;
            variance = sumSquares - (sum*sum) / w;
            int i4 = lowerClassLimit-1;
            if(i4 != 0)
            {
                for(int j = 2; j < nClasses+1; j++)
                {
                    if(varianceCombinations[l][j] >= (variance + varianceCombinations[i4][j-1])){
                        lowerClassLimits[l][j] = lowerClassLimit;
                        varianceCombinations[l][j] = variance + varianceCombinations[i4][j-1];
                    }
                }
            }
        }
        lowerClassLimits[l][1] = 1.0;
        varianceCombinations[l][1] = variance;
    }
    return std::tuple<std::vector<std::vector<double>>, std::vector<std::vector<double>>>{lowerClassLimits, varianceCombinations};
}

std::vector<double> getJenksBreaks(std::vector<double> data, std::vector<std::vector<double>> lowerClassLimits, int nClasses){
    int k = data.size() - 1;
    std::vector<double> kClass(nClasses+1, 0.0);
    int countNum = nClasses;
    kClass[0] = data[0];
    kClass[nClasses] = data[k];
    while(countNum > 1){
        int elt = lowerClassLimits[k][countNum]-2;
        kClass[countNum-1] = data[elt];
        k = lowerClassLimits[k][countNum]-1;
        countNum -= 1;
    }
    return kClass;
}

// TODO: preallocate memory for clusters and access with [] instead of pushing back
std::vector<std::vector<double>> separateData(std::vector<double> data, std::vector<double> boundaries){
    std::vector<std::vector<double>> clusters = {};
    std::vector<double> cluster = {};

    boundaries.erase(boundaries.begin());
    int i = 0;

    for(double& d : data){
        if(d <= boundaries[i])
            cluster.push_back(d);
        else{
            clusters.push_back(cluster);
            cluster = {d};
            i += 1;
        }
    }

    clusters.push_back(cluster);
    return clusters;

}

std::tuple<std::vector<std::vector<double>>, std::vector<double>> jenks(std::vector<double> data, int nClasses){
//    assert(nClasses > data.size());
    if (nClasses > data.size())
        std::cout << nClasses << " > " << data.size() << std::endl;
    std::sort(data.begin(), data.end());
    std::tuple<std::vector<std::vector<double>>, std::vector<std::vector<double>>> results = jenksMatrices(data, nClasses);
    std::vector<std::vector<double>> lowerClassLimits = std::get<0>(results);
    std::vector<std::vector<double>> x = std::get<1>(results);
    std::vector<double> boundaries = getJenksBreaks(data, lowerClassLimits, nClasses);
    return std::tuple<std::vector<std::vector<double>>, std::vector<double>>{separateData(data, boundaries), boundaries};
}
